torch/jit/__init__.py

```diff
@@ -456,7 +457,163 @@ def trace(*args, **kwargs):
+        raise RuntimeError("Only parameters and buffers of compiled modules can be re-assigned.")
+
+    def __delattr__(self, name):
+        raise RuntimeError("Deleting attributes of TracedModules isn't supported")
+
+    def __getstate__(self):
+        raise RuntimeError("TracedModules aren't picklable")
+
+    def __call__(self, *args):
+        try:
+            return self._executor(*args)
+        except AttributeError:
+            raise RuntimeError("Only the top-level compiled module can be called")
+
+    def _recompute_captures(self):
+        if self._is_root:
+            self._executor.set_captures(*self.state_dict().values())
+        else:
+            root = self._root()
+            if root is None:
+                raise RuntimeError("Submodules of TracedModule can't work without keeping "
+                                   "the main one in scope")
+            root._recompute_captures()
+
+    def _apply(self, fn):
+        if not self._is_root:
+            raise RuntimeError("Only the top-level compiled module supports type casts")
+        for module in itertools.chain((self,), self.modules()):
+            for param in module._parameters.values():
+                param.data = fn(param.data)
+                if param._grad is not None:
+                    param._grad.data = fn(param._grad.data)
+
+            for key, buf in module._buffers.items():
+                module._buffers[key] = fn(buf)
+        self._recompute_captures()
+        return self
+
+    def register_parameter(self, name, param):
+        if name not in self._parameters:
+            raise RuntimeError("Can't add new parameters to TracedModules")
+        if param is None:
+            raise RuntimeError("Can't set parameters to None in TracedModules")
+        super(TracedModule, self).register_parameter(name, param)
+
+    def load_state_dict(self, state):
+        super(TracedModule, self).load_state_dict(state)
+        # NB: this is not strictly necessary, because load_state_dict is copying, but
+        # that's an implementation detail that I don't want to depend on
+        self._recompute_captures()
+
+    def _get_name(self):
+        return 'TracedModule[' + self._orig_class.__name__ + ']'
+
+    def __repr__(self):
+        # Subclasses usually print some extra info like num_features, but this needs
+        # to have them saved as attributes, which TracedModules lack.
+        return Module.__repr__(self)
+
+def _get_methods(cls):
+    import inspect
+    # In Python 3 unbound methods are functions, but in Python 2 they are methods
+    return inspect.getmembers(cls, predicate=lambda x: inspect.isfunction(x) or inspect.ismethod(x))
+
+# NOTE: only need include those that aren't implemented above
+_compiled_methods_whitelist = {
+    'cpu', 'cuda', 'double', 'float', 'half', 'modules', 'named_children',
+    'named_modules', 'named_parameters', 'parameters', 'state_dict', 'type',
+    'zero_grad',
+}
+def _make_fail(name):
```

Comment:
I think this stuff is confusing. I would expect to just get attribute not found errors. This is why I think that traced modules should not pretend to be subclasses of the original module type. There are not sane semantics where they can be considered subclasses and then we need to go through and add these overloads to try to make them behave.

Reply:
If we want to make traced modules subclasses of `nn.Module` then there's no way to make them raise attribute errors (without changing `_make_fail` to raise them). If we don't define them here, they will simply dispatch to the superclass, and we want to avoid it. It has nothing to do with subclassing the concrete class, it only compares `TracedModule` to `nn.Module`, and is run once, not every time a subclass is created.

Reply:
How many methods is this actually blacklisting? If it is less than 10 or so, we should just write the implementations directly.  It only makes sense to resort to python class hacking when the other solutions become unreadable. 

If there are a lot of methods, we should consider refactoring nn.Module into a base class that has the barebones functionality, and another class that implements the functionality that a standard python module needs. Eventually we are going to have C++ classes that Peter/Greg are working on that should behave like modules, and they will also need to provide their own ways of implementing some module methods.

Reply:
I could do that, but this way is more future proof. Remember that we can always extend the `nn.Module` API, and I don't want to have to think if this doesn't break the JIT in any way. This code ensures that it will blacklist everything we don't want by default, and is thus safer.

I don't think adding a base class is a good idea (at least just yet). JIT should complement `nn`, not require large refactors. There's no such thing as a "barebones" module, and it doesn't make sense in context of `nn`.

Reply:
The way to keep nn.Module future proof is do define its abstract API in a way that TracedModule can implement, rather than have TracedModule hack around the default implementation of module. If there was an abstract base class that showed which methods needed to be filled in, then inheriting from that would give you the correct futureproof behavior, and would be idiomatic python. We need to do something like this anyway to support C++ modules, and CompiledModules. If we don't have an abstract list of functionality these modules must provide then I don't know how we will implement them correctly.

Reply:
We do have a list now, but it's not guaranteed to be frozen. If someone goes on to refactor `nn.Module` as an ABC that will let us catch these errors then fine, but until then I think this is strictly superior to letting things blow up when people try to use `TracedModule`s, just because we forgot to update this part of the code. It's not even very complicated, it's just registering methods that raise errors.

Is there any doc/conversation on C++ modules? I'm interested in this because I did some related work for DataParallel but I haven't heard anything about it.

Reply:
Also, I don't think that basing the module API on the JIT constraints is a good idea. Modules have a lot of useful things that are disabled just by the fact that `TracedModule` needs to be pretty much immutable. Embedding this requirement into all modules would be harmful.